geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java

```diff
@@ -3453,6 +3448,12 @@ public ClusterDistributionManager getDM() {
       return dm;
     }
 
+    @Override
+    public void saveConfig() {
```

Comment:
Rather than having the DM know so much detail about the Cache and that it needs to save a cache.xml, wouldn't it be better to design reconnect to use a listener notification mechanism to notify all Geode components that a reconnect is occurring... maybe beforeReconnect and afterReconnect and then allow the components to decide for themselves how to react (ie, save the cache.xml)?

Reply:
I agree. Actually, I really wanted to put this logic inside the existing membershipFailure event. But looking at where saveConfig is called currently, it doesn't really seem to match up with starting a reconnect in all cases. That may be a bug, I'm not sure.

For this change I was just trying to fix the circular dependencies and Cache.getInstance calls inside membership without changing the behavior. If it's ok with you, I'd like to change the reconnect API between Cache/DM and the membership layer in a later commit.

BTW - I really would prefer that *nothing* knows about reconnect at all except for maybe some small wrapper (even a wrapper shell script?). We've hit too many bugs because we have special "reconnect" logic that is somehow different than "startup" logic.